Changes in module 'nf-core/krakentools/extractkrakenreads'
--- modules/nf-core/krakentools/extractkrakenreads/main.nf
+++ modules/nf-core/krakentools/extractkrakenreads/main.nf
@@ -45,4 +45,28 @@
         extract_kraken_reads.py: ${VERSION}
     END_VERSIONS
     """
+
+    stub:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def extension = args.contains("--fastq-output") ? "fastq" : "fasta"
+    def input_reads_command = meta.single_end ? "-s $classified_reads_fastq" : "-s1 ${classified_reads_fastq[0]} -s2 ${classified_reads_fastq[1]}"
+    def output_reads_command = meta.single_end ? "-o ${prefix}.extracted_kraken2_read.${extension}" : "-o ${prefix}.extracted_kraken2_read_1.${extension} -o2 ${prefix}.extracted_kraken2_read_2.${extension}"
+    def report_option = report ? "-r ${report}" : ""
+    def VERSION = '1.2' // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
+
+    """
+    if [ "$meta.single_end" == "true" ]; 
+    then
+        touch ${prefix}.extracted_kraken2_read.${extension}
+    else
+        touch ${prefix}.extracted_kraken2_read_1.${extension}
+        touch ${prefix}.extracted_kraken2_read_2.${extension}
+    fi
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        extract_kraken_reads.py: ${VERSION}
+    END_VERSIONS
+    """
 }

************************************************************
