# rki-mf1/igsmp: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

  > Andrews, S. (2010). FastQC: A Quality Control Tool for High Throughput Sequence Data [Online].

- [MultiQC](https://pubmed.ncbi.nlm.nih.gov/27312411/)

  > Ewels P, Magnusson M, Lundin S, Käller M. MultiQC: summarize analysis results for multiple tools and samples in a single report. Bioinformatics. 2016 Oct 1;32(19):3047-8. doi: 10.1093/bioinformatics/btw354. Epub 2016 Jun 16. PubMed PMID: 27312411; PubMed Central PMCID: PMC5039924.

- [BCFtools](https://pubmed.ncbi.nlm.nih.gov/33590861/)

  > Danecek P, Bonfield JK, Liddle J, Marshall J, Ohan V, Pollard MO, Whitwham A, Keane T, McCarthy SA, Davies RM, Li H. Twelve years of SAMtools and BCFtools [Internet]. GigaScience. Oxford University Press (OUP); 2021. Available from: http://dx.doi.org/10.1093/gigascience/giab008

- [bwa](https://arxiv.org/abs/1303.3997)

  > Li H. Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM [Internet]. arXiv; 2013. Available from: https://arxiv.org/abs/1303.3997 

- [fastp](https://onlinelibrary.wiley.com/doi/10.1002/imt2.107)

  > Chen S. Ultrafast one‐pass FASTQ data preprocessing, quality control, and deduplication using fastp [Internet]. iMeta. Wiley; 2023. Available from: http://dx.doi.org/10.1002/imt2.107

- [kraken2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1891-0)

  > Wood DE, Lu J, Langmead B. Improved metagenomic analysis with Kraken 2 [Internet]. Genome Biology. Springer Science and Business Media LLC; 2019. Available from: http://dx.doi.org/10.1186/s13059-019-1891-0 

- [krakentools](https://www.nature.com/articles/s41596-022-00738-y)

  > Lu J, Rincon N, Wood DE, Breitwieser FP, Pockrandt C, Langmead B, Salzberg SL, Steinegger M. Metagenome analysis using the Kraken software suite [Internet]. Nature Protocols. Springer Science and Business Media LLC; 2022. p. 2815–2839. Available from: http://dx.doi.org/10.1038/s41596-022-00738-y 

- [lofreq](https://pubmed.ncbi.nlm.nih.gov/23066108/)

  > Wilm A, Aw PPK, Bertrand D, Yeo GHT, Ong SH, Wong CH, Khor CC, Petric R, Hibberd ML, Nagarajan N. LoFreq: a sequence-quality aware, ultra-sensitive variant caller for uncovering cell-population heterogeneity from high-throughput sequencing datasets [Internet]. Nucleic Acids Research. Oxford University Press (OUP); 2012. p. 11189–11201. Available from: http://dx.doi.org/10.1093/nar/gks918 

- [Picard](https://github.com/broadinstitute/picard)

  > Picard toolkit [Internet], Broad Institute. Available from: https://github.com/broadinstitute/picard

- [samtools](https://academic.oup.com/gigascience/article/10/2/giab008/6137722

  > Danecek P, Bonfield JK, Liddle J, Marshall J, Ohan V, Pollard MO, Whitwham A, Keane T, McCarthy SA, Davies RM, Li H. Twelve years of SAMtools and BCFtools [Internet]. GigaScience. Oxford University Press (OUP); 2021. Available from: http://dx.doi.org/10.1093/gigascience/giab008 

- [tabix](https://academic.oup.com/bioinformatics/article/27/5/718/262743)

  > Li H. Tabix: fast retrieval of sequence features from generic TAB-delimited files [Internet]. Bioinformatics. Oxford University Press (OUP); 2011. p. 718–719. Available from: http://dx.doi.org/10.1093/bioinformatics/btq671 

- [bedtools](https://academic.oup.com/bioinformatics/article/26/6/841/244688)

  > Quinlan AR, Hall IM. BEDTools: a flexible suite of utilities for comparing genomic features [Internet]. Bioinformatics. Oxford University Press (OUP); 2010. p. 841–842. Available from: http://dx.doi.org/10.1093/bioinformatics/btq033 

- [Biopython](https://academic.oup.com/bioinformatics/article/25/11/1422/330687)

  > Cock PJA, Antao T, Chang JT, Chapman BA, Cox CJ, Dalke A, Friedberg I, Hamelryck T, Kauff F, Wilczynski B, de Hoon MJL. Biopython: freely available Python tools for computational molecular biology and bioinformatics [Internet]. Bioinformatics. Oxford University Press (OUP); 2009. p. 1422–1423. Available from: http://dx.doi.org/10.1093/bioinformatics/btp163 

- [PyVCF](https://pyvcf.readthedocs.io/en/latest/)


## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

  > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

  > Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2. doi: 10.5555/2600239.2600241.

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)

  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.
